parley/src/layout/cursor.rs

```diff
@@ -183,4 +233,250 @@ impl CursorPath {
     pub fn cluster<'a, B: Brush>(&self, layout: &'a Layout<B>) -> Option<Cluster<'a, B>> {
         self.run(layout)?.get(self.cluster_index)
     }
+
+    pub fn visual_line<'a, B: Brush>(&self, layout: &'a Layout<B>) -> Option<Line<'a, B>> {
+        layout.get(self.visual_line_index)
+    }
+}
+
+/// Returns a point that is falls within the vertical bounds of the given line.
+fn line_y<B: Brush>(line: &Line<B>) -> f32 {
+    line.metrics().baseline - line.metrics().ascent * 0.5
+}
+
+#[derive(Copy, Clone, PartialEq, Default, Debug)]
+pub struct Selection {
+    anchor: Cursor,
+    focus: Cursor,
+    h_pos: Option<f32>,
+}
+
+impl From<Cursor> for Selection {
+    fn from(value: Cursor) -> Self {
+        Self {
+            anchor: value,
+            focus: value,
+            h_pos: None,
+        }
+    }
+}
+
+impl Selection {
+    pub fn from_point<B: Brush>(layout: &Layout<B>, x: f32, y: f32) -> Self {
+        Cursor::from_point(layout, x, y).into()
+    }
+
+    pub fn from_byte_index<B: Brush>(layout: &Layout<B>, index: usize) -> Self {
+        Cursor::from_byte_index(layout, index).into()
+    }
+
+    pub fn anchor(&self) -> &Cursor {
+        &self.anchor
+    }
+
+    pub fn focus(&self) -> &Cursor {
+        &self.focus
+    }
+
+    pub fn is_collapsed(&self) -> bool {
+        self.anchor.insert_point == self.focus.insert_point
+    }
+
+    pub fn text_range(&self) -> Range<usize> {
+        if self.anchor.text_start < self.focus.text_start {
+            self.anchor.text_start..self.focus.text_start
+        } else {
+            self.focus.text_start..self.anchor.text_start
+        }
+    }
+
+    /// Returns the index where text should be inserted based on this
+    /// selection.
+    pub fn insertion_index(&self) -> usize {
+        self.focus.text_start
+    }
+
+    pub fn collapse(&self) -> Self {
+        Self {
+            anchor: self.focus,
+            focus: self.focus,
+            h_pos: self.h_pos,
+        }
+    }
+
+    pub fn extend_to_point<B: Brush>(&self, layout: &Layout<B>, x: f32, y: f32) -> Self {
+        Self {
+            anchor: self.anchor,
+            focus: Cursor::from_point(layout, x, y),
+            h_pos: None,
+        }
+    }
+
+    pub fn next_logical<B: Brush>(&self, layout: &Layout<B>, extend: bool) -> Self {
+        self.maybe_extend(Cursor::from_byte_index(layout, self.focus.text_end), extend)
```

Comment:
This doesn't quite work for emergency breaks. Pressing right at:
```
aaaaaaaaaaaaaaaaaaaaaa<|>a[emergency break]
aaaaaa
```
should give:
```
aaaaaaaaaaaaaaaaaaaaaaa<|>[emergency break]
aaaaaa
```
but actually gives:
```
aaaaaaaaaaaaaaaaaaaaaaa[emergency break]
<|>aaaaaa
```

A similar thing happens with bidi text (and would also happen with differently styled rich text).

Reply:
My understanding is that right now Parley doesn't keep track of cursor affinity. This is not unheard of (chrome doesn't have affinity either) but I also strongly prefer keeping affinity for soft breaks. It's also the behaviors of most text editors.

Reply:
I was under the impression that this logic already correctly has state for affinity, as discussed in https://github.com/linebender/parley/pull/106#discussion_r1720081728

In the example I have given, end and home work as expected for me, the issue is with the left and right keys.

Reply:
I do have this working for soft line breaks but not emergency breaks (yet). This new behavior matches Firefox where moving right at the end of the line places the cursor at the beginning of the next (rather than skipping the first cluster). Note that this effectively requires two key presses to move between characters so affinity doesn’t handle this naturally.